### ISSUE TITLE
Bug: [no-floating-promises] allows unhandled Promise rejections

### ISSUE THREAD

**Issue by janparisek** [typescript-eslint/typescript-eslint#7685]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.ts&code=G4ewlgJgBAFAhgZwJ4DsDGsCUUC8A%2BKAbwFgAoKKNEFBAFygAcAnEAWzAQFNcoVOB3KAAUW7LjBhNOCEABtgnADRQpAK05pa2fERWd1mmACJgcWQFdOR7AF9MAbjIU4-OGHrM2HTo9J2YDmRkAPTBUNKyYCi0ALQQHHAARrKcMXwAHrGRfFAAArRIDNJoTGAMsRFRtMEoIDHmKOZcEDGmTAhkVDQeot4ATDx8giJe4pLScgrKahpauASEega0xqYWVraYTlAAdLQAFpwoEtoLUHa%2BQA&eslintrc=N4KABGBECmAeAu0B2ATAzpAXGA2uCU0aANgJZLyYBO0AxgPYC2jyK0KkANPhJAA7EArgHNymAALwAnnyK0qpPvAC0RMhQD0aeAtorps5bQAWdANbtI%2BALrdefAIZU00KliiSZchUtUly8BqOzq5c%2BPxOLlQA8kqk9EgY2KAEUHxU9ABWdPDukPBoDEgAZqTCAHSZaAlWEAC%2BdmlCoonueKmQnrKFPipqAbVgtuEZ9LnYOoLQIHUgQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkeAhpgObp4VTokBMADF15AA04GpHiwAJgFcM7SEUnoAjADYhI2pwCykmegCSRPFliJ2eTFPTCaUAGaxM0dAGFiJZCWNGX9D0SZDADE0dAA5enhOc0trDSgvTGRoNkoLKxtREgBrZEQAGWQAIxcAC3RobJirEQBfEHqgA&tokens=false

### Repro Code

```typescript
void (async () => {
  const promise = new Promise((resolve, reject) => { reject("value") });
  await promise;
})();

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const promise2 = new Promise((resolve, reject) => { reject("value") })
  .then(() => { });
```

### ESLint Config

```javascript
/* eslint-env node */
module.exports = {
  extends: [
    "eslint:recommended",
    "plugin:@typescript-eslint/strict-type-checked",
  ],
  parser: "@typescript-eslint/parser",
  parserOptions: {
    project: "tsconfig.json",
  },
  plugins: ["@typescript-eslint"],
  root: true,
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "target": "ES2022",
    "module": "Node16",
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "skipLibCheck": true
  }
}
```

### Expected Result

3rd line should cause `no-floating-promises` error.
7th/8th line should cause `no-floating-promises` error.

### Actual Result

No error detected when linting.

`UnhandledPromiseRejection` during run time.

### Additional Info

Those are two examples I came across while coding. Both seem to be different cases, but I didn't want to open two different issues.

**Comment by bradzacher** [comment]
> 3rd line should cause no-floating-promises error.

Why should it be an error?
The promise is awaited so the exception is deferred to the parent.
And then the parent has been explicitly marked with `void` which signals to the rule that its promise is purposely floating.

> 7th/8th line should cause no-floating-promises error.

promises assigned to variables are not floating promises.
it is assumed that the variable is correctly used.

the fact that you've left it unused is a different error which is caught by the recommended `no-unused-vars` rule.

**Comment by janparisek** [comment]
I see.

Thanks for taking your time.
